### DISCUSSION TITLE
Python3 unit tests are failing when run by the GitHub actions runner, but they pass when I run them locally.

### DISCUSSION THREAD

**Discussion by ZathrasXI** [community/community#49808] [Actions] [answered]
### Select Topic Area

Question

### Body

This is my first time using GitHub actions, and attempting any sort of CI development work.

I have a [basic spreadsheet reader app](https://github.com/ZathrasXI/JointSpendingCalclutor), written in Python. There are 12 unit tests, all of them pass when I run them on my Ubuntu laptop. The last 6 of them are failing when I run them in GitHub actions. Some of these tests read, write, and create csv files - could there be an issue with permissions in the runner?

Apart from this obstacle I have found the instructions, and setup for GitHub actions to be easy!

**Comment by Tinniqqss** [comment]
> ### Select Topic Area
> 
> Question
> 
> ### Body
> 
> This is my first time using GitHub actions, and attempting any sort of CI development work.
> 
> I have a [basic spreadsheet reader app](https://github.com/ZathrasXI/JointSpendingCalclutor), written in Python. There are 12 unit tests, all of them pass when I run them on my Ubuntu laptop. The last 6 of them are failing when I run them in GitHub actions. Some of these tests read, write, and create csv files - could there be an issue with permissions in the runner?
> 
> Apart from this obstacle I have found the instructions, and setup for GitHub actions to be easy!
>

**Reply by ZathrasXI** [reply]
What was your answer?

**Comment by airtower-luna** [comment, marked as answer]
The [test module](https://github.com/ZathrasXI/JointSpendingCalclutor/blob/289fa8df58daaaf9732e81967c61b2d5c1f7725b/test_wow.py) is written using `unittest`, but uses plain `assert`s. That means you only see `AssertionError`s in the `unittest` output, not *how* the assertions failed. You have two options there:

1. If you want to stick with plain `unittest`, use the appropriate [assertion methods](https://docs.python.org/3.11/library/unittest.html#assert-methods) instead of the `assert` keyword.
2. What I'd do is install [`pytest`](https://docs.pytest.org/en/7.2.x/), and run your tests with it instead of `unittest`. `pytest` does its own handling of assertions and will show you why they failed. `pytest` can run tests written for `unittest`, so you don't have to rewrite your tests (though you could, if you want to use other `pytest` features).

If you use either of those options debugging should get a lot easier at least. :slightly_smiling_face:

**Reply by ZathrasXI** [reply]
Thank you! I am experimenting with PyTest now :) I agree, the detailed output is making this easier to debug! Have a good day.

**Reply by ZathrasXI** [reply]
All resolved! Thank you again @airtower-luna. The unit tests are now passing in the pipeline :D, I'm enjoying PyTest a lot - especially the `fixtures`

**Reply by airtower-luna** [reply]
Excellent, glad I could help! :slightly_smiling_face:
